### ISSUE TITLE
Telescope pattern from paper not working in turnstile-core

### ISSUE THREAD

**Issue by wilbowma** [stchang/macrotypes#95]
I tried to use a pattern that appears essentially verbatim in the paper, but it's not working:
```
#lang turnstile

(define-base-type Type-)

(define-syntax Type
  (syntax-parser
    [:id
     (syntax-property ': #'(Type-) #'Type)]))


(struct ->- (types))
(define-typed-syntax (-> [x : τ₁] ... τ₂) ≫
  [[x ≫ x- : τ₁ ≫ τ₁- ⇐ Type] ... ⊢ [τ₂ ≫ τ₂- ⇐ Type]]
  ------
  [⊢ (->- (λ (x- ...) τ₁- ... τ₂-))])

(define-typed-syntax (->/i [x : τ₁] ... τ₂) ≫
  [[x ≫ x- : τ₁ ≫ τ₁- ⇒ ~Type] ... ⊢ [τ₂ ≫ τ₂- ⇒ ~Type]]
  ------
  [⊢ (->- (λ (x- ...) τ₁- ... τ₂-))])
```
This is essentially the same as the line in Figure 7 for how to do this "right". However, I get:
```
syntax/loc: no pattern variables before ellipsis in template
;   at: τ₁-
```
at lines 15 and 20.

**Comment by wilbowma** [comment]
Perhaps related, perhaps not:
```
#lang turnstile

(define-base-type Type-)

(define-syntax Type
  (syntax-parser
    [:id
     (syntax-property ': #'(Type-) #'Type)]))


(struct ->- (types))
(define-typed-syntax (-> [x : τ₁] ... τ₂) ≫
  [[x ≫ x- : τ₁ ≫ τ₁-] ... ⊢ [τ₂ ≫ τ₂- ⇒ ~Type-]]
  ------
  [⊢ (->- (λ (x- ...) τ₂-))])

(-> [x : Type] [y : Type] Type)
```

This gives a different error, complaining about the `≫` between ` τ₁` and `τ₁-`. The displayed pattern also makes no sense to me.

```
; /tmp/<pkgs>/turnstile-lib/turnstile/turnstile.rkt:345:33: ?: unexpected term
;   at: ≫
;   in: ((x : Type ≫ τ₁-) (y : Type ≫ τ₁-))
```

**Comment by stchang** [comment]
Oops, it's an ambiguous parsing problem, because non-telebind lhs context entries can have multiple "types", keyed on any arbitrary identifier. Eg, it thinks `x` has `:` "type" = `τ₁`, and `≫` "type" = `τ₁-`. The existing test cases must have had non-empty telebinds on the rhs of the `⊢` that disambiguated the parsing.

**Comment by wilbowma** [comment]
I tried e78a968c4e0cfcab1c73efccc127f0ca31eceb80 and now the second example works (after I fixed some bugs in it).

```
#lang turnstile

(define-base-type Type-)

(define-syntax Type
  (syntax-parser
    [:id
     (syntax-property #'Type- ': #'Type)]))

(begin-for-syntax
  (define-syntax ~Type
    (pattern-expander
     (syntax-parser
       [tmp:id
        #`(~or (~literal Type) ~Type-)]))))


(struct ->- (types))
(define-typed-syntax (-> [x : τ₁] ... τ₂) ≫
  [[x ≫ x- : τ₁ ≫ τ₁- ⇒ _] ... ⊢ [τ₂ ≫ τ₂- ⇒ ~Type]]
  ------
  [⊢ (->- (λ (x- ...) τ₂-)) ⇒ Type])

(-> [x : Type] [y : Type] Type)
```

The first example has problems in check mode:
```
#lang turnstile

(define-base-type Type-)

(define-syntax Type
  (syntax-parser
    [:id
     (syntax-property #'Type- ': #'Type)]))

(begin-for-syntax
  (define-syntax ~Type
    (pattern-expander
     (syntax-parser
       [tmp:id
        #`(~or (~literal Type) ~Type-)]))))

(struct ->- (types))
(define-typed-syntax (-> [x : τ₁] ... τ₂) ≫
  [[x ≫ x- : τ₁ ≫ τ₁- ⇐ Type] ... ⊢ [τ₂ ≫ τ₂- ⇐ Type]]
  ------
  [⊢ (->- (λ (x- ...) τ₁- ... τ₂-))])

(define-typed-syntax (->/i [x : τ₁] ... τ₂) ≫
  [[x ≫ x- : τ₁ ≫ τ₁- ⇒ ~Type] ... ⊢ [τ₂ ≫ τ₂- ⇒ ~Type]]
  ------
  [⊢ (->- (λ (x- ...) τ₁- ... τ₂-))])
```
```
; /tmp/meow.rkt:20:24: syntax: no pattern variables before ellipsis in template
;   at: Type
;   in: (syntax (Type ...))
; [Due to errors, REPL is just module language, requires, and stub definitions]
```

**Comment by wilbowma** [comment]
Similar example which causes the same problems with the tag ':
```
#lang turnstile

(define-base-type Type-)

(define-syntax Type
  (syntax-parser
    [:id
     (syntax-property #'Type- ': #'Type)]))

(begin-for-syntax
  (define-syntax ~Type
    (pattern-expander
     (syntax-parser
       [tmp:id
        #`(~or (~literal Type) ~Type-)]))))

(struct ->- (types))
(define-typed-syntax (-> [x (~datum :) τ₁] ... τ₂) ≫
  [[x ≫ x- : τ₁ ≫ τ₁- ⇐ Type] ... ⊢ [τ₂ ≫ τ₂- ⇐ Type]]
  ------
  [⊢ (->- (λ (x- ...) τ₁- ... τ₂-))])

(define-typed-syntax (->/i [x (~datum :) τ₁] ... τ₂) ≫
  [[x ≫ x- : τ₁ ≫ τ₁- ⇒ ~Type] ... ⊢ [τ₂ ≫ τ₂- ⇒ ~Type]]
  ------
  [⊢ (->- (λ (x- ...) τ₁- ... τ₂-))])
```
```
; /tmp/meow2.rkt:19:11: syntax: no pattern variables before ellipsis in template
;   at: :
;   in: (syntax (: ...))

```

**Comment by stchang** [comment]
I have a fix for this. The ellipsis pattern is handled wrong. Just testing some more before pushing.

**Comment by stchang** [comment]
Another partial fix in 5bc86b10a77b17789c3ae44c7d357e031eed242c

All except https://github.com/stchang/macrotypes/issues/95#issuecomment-517415325 should be working now

**Comment by stchang** [comment]
This last commit should fix all the problems 35c0570d1f94c4bef17e1ac2f1ff06936580b2d5

**Comment by wilbowma** [comment]
This two still fails:

1.
```racket
#lang turnstile

(define-base-type Type-)

(define-syntax Type
  (syntax-parser
    [:id
     (syntax-property #'Type- ': #'Type)]))

(begin-for-syntax
  (define-syntax ~Type
    (pattern-expander
     (syntax-parser
       [tmp:id
        #`(~or (~literal Type) ~Type-)]))))

(struct ->- (types))

(define-typed-syntax (->/i [x (~datum :) τ₁] ... τ₂) ≫
  [[x ≫ x- : τ₁ ≫ τ₁- ⇒ ~Type] ... ⊢ [τ₂ ≫ τ₂- ⇒ ~Type]]
  ------
  [⊢ (->- (λ (x- ...) τ₁- ... τ₂-))])
```

```
Welcome to Racket v7.1.0.10.
; <pkgs>/turnstile-lib/turnstile/turnstile.rkt: ~Type: expected identifier
;   at: (~Type ...)
;   in: (~Type ...)
```

2.
```racket
#lang turnstile

(define-base-type Type-)

(define-syntax Type
  (syntax-parser
    [:id
     (syntax-property #'Type- ': #'Type)]))

(begin-for-syntax
  (define-syntax ~Type
    (pattern-expander
     (syntax-parser
       [tmp:id
        #`(~or (~literal Type) ~Type-)]))))

(struct ->- (types))

(define-typed-syntax (->/i [x (~datum :) τ₁] ... τ₂) ≫
  [[x ≫ x- : τ₁ ≫ τ₁- ⇒ _] ... ⊢ [τ₂ ≫ τ₂- ⇒ _]]
  ------
  [⊢ (->- (λ (x- ...) τ₁- ... τ₂-))])
```
```
; /tmp/meow.rkt:20:11: syntax: no pattern variables before ellipsis in template
;   at: :
;   in: (syntax (: ...))
```
